### ISSUE TITLE
Wasm target architecture for native image

### ISSUE THREAD

**Issue by jespertheend** [oracle/graal#3391]
**Feature request**

**Is your feature request related to a problem? Please describe.**
I'm trying to port closure compiler to wasm so that it can run from within a browser.

**Describe the solution you'd like.**
Closure compiler already has build support for native image. So I was hoping to be able to use the LLVM backend but it seems like adding a new target architecture is a lot more work than I had expected.

**Describe who do you think will benefit the most.**
Anyone using GraalVM to build native image versions of their app would benefit from this, as they can now port their application to the web easily. Java developers who want to write a new application from the web could also benefit from this, they can now start developing in a language that they are already familiar with.

**Describe alternatives you've considered.**
I have tried the following alternatives:
- [CheerjP](https://leaningtech.com/cheerpj/), works and runs closure compiler in a browser successfully, but the runtime is rather bulky, which needs to be downloaded first before you can run jar files. I'm also not sure about the performance. Wasm could be faster. Though I don't have any evidence to back this up.
- [TeaVM](https://github.com/konsoletyper/teavm), could work as well. Though I haven't tried this yet to be honest. Their docs seem a bit daunting.
- [JWebAssembly](https://github.com/i-net-software/JWebAssembly) might work but it is currently not production ready and still missing some crucial features.
- [Bytecoder](https://github.com/mirkosertic/Bytecoder) doesn't seem to work for my use case since it doesn't support reflection. Something that closure compiler depends on.

**Additional context.**
The links in [LLVM Backend for Native Image](https://github.com/oracle/graal/blob/89e4cfc7aeea69970b60c64cd075ceb2a104e864/substratevm/LLVMBackend.md) seem useful for implementing a new target.

**Express whether you'd like to help contributing this feature**
I'd love to, though if this feature request requires a similar change to the changes in [[AArch64][Statepoints] Statepoint support for AArch64.](https://reviews.llvm.org/D66012), I don't think I'm the suitable person to do this. I was hoping it would only require a few flags to be set on llvm but it seems like that's not the case.

**Comment by mcraj017** [comment]
@jespertheend thanks for the feature request, we are constantly adding new support to native-image and if we find it valuable to the community then surely we will add it in our future releases.

**Comment by mcraj017** [comment]
@christianwimmer please advise

**Comment by christianwimmer** [comment]
We have thought about a WASM backend, but it is not on our list of features to implement in the near future.
One of the biggest problems is the missing GC support in WASM: neither the GC of the underlying WASM/JS VM is exposed, nor stack walking is possible to include the existing SVM GC.

**Comment by jespertheend** [comment]
Ah yeah makes sense. Thanks for the info!
Would the wasm gc proposal make this any easier?

**Comment by christianwimmer** [comment]
> Would the wasm gc proposal make this any easier?
Yes, once WASM has a proper GC interface things are getting much more feasible.

**Comment by jespertheend** [comment]
Ah cool! I stay on the lookout for that then. Thanks for the info.

**Comment by jimmywarting** [comment]
Don't seem to be the only one wanting to port closure compiler to wasm 😉 
subscribing to this issue. 

ping me @jespertheend if you ever release a wasm version of closure compiler :)

**Comment by jespertheend** [comment]
@jimmywarting I'm actually about to ditch closure compiler in favour of terser to be honest. From my limited testing, builds seem to be slightly larger, but that seems worth it considering that it can run in a browser and generally seems to have less complexity.
You might have some luck with https://github.com/mirkosertic/Bytecoder but some features seem to still be missing for it to successfully build closure compiler to wasm.

**Comment by sdeleuze** [comment]
@christianwimmer Potentially interesting related links:
 - [.Net notes about why they can't leverage GC proposal yet](https://github.com/WebAssembly/gc/issues/77)
 - [LLVM and clang progresses on GC proposal support](https://github.com/WebAssembly/gc/issues/77#issuecomment-1195488265)
 - [Draft WASI support in TeaVM](https://github.com/konsoletyper/teavm/pull/610)
 - [J2CL](https://github.com/google/j2cl) which currently provide the most advanced Java to Wasm support (and leverages the GC proposal)

**Comment by mirkosertic** [comment]
Hi!

Author of https://github.com/mirkosertic/Bytecoder here. I can confirm that native garbage collection support is an issue in WebAssembly. The current Wasm GC proposal is in "proposed" status, as seen here: https://chromestatus.com/feature/6062715726462976. However, Google's J2CL Wasm backend seems to rely on this proposal, so I am really not sure when we can expect the final release and support in Chromium at least. Till this point, every Wasm backend for JVM bytecode needs to provide its own implementation for garbage collection. And implementing a fast and reliable garbage collector can be really tricky, and it can become a nightmare and impossible if you want to prevent stop-the-world during GC runs in a Wasm in a single threaded environment. Maybe we can get around that with reference counting, but performance is a very important part of it. Good escape analysis will remove some of the heap allocations, but in most real life applications a lot of stack allocation will remain.

But when the time comes and the Wasm GC proposal is final, it would be good to know what might be a feasible way to use the Graal native image compiler with Wasm. The LLVM backend path might work, but it is for sure not the fastest way. So what I am searching is a kind of documentation about how to provide a custom backend to Graal, or a kind of "embedded" API to run Graal, retrieve the IR (Graphs) and use them for a custom backend for code generation. Something that can be completely based on JVM, without further toolchains like LLVM and co.

**Comment by sdeleuze** [comment]
GC proposal made a lot of progresses in 2022 and moved to [phase 3](https://github.com/WebAssembly/proposals). We can't be sure when it will move to phase 4 but could be possible in 2023.

Implementations in other browsers has started:
 - [FireFox (SpiderMonkey)](https://bugzilla.mozilla.org/show_bug.cgi?id=1774825)
 - [Safari & WebKit (JavaScriptCore)](https://bugs.webkit.org/show_bug.cgi?id=247394)

**Comment by InSuperposition** [comment]
With Docker + WASM in beta, seems like this targeting WASM would have a few more benefits for containers. Thanks for the great work!

https://www.docker.com/blog/docker-wasm-technical-preview/

https://docs.docker.com/desktop/wasm/

**Comment by MTyson** [comment]
Hello, I'm writing an article for InfoWorld about Java and WASM.  native-image --features=wasm sure would be nice.  Could anyone comment on the status or any thoughts?

**Comment by tenaf0** [comment]
There has been quite a lot of movement around the WASM GC implementation, and it is available by default on chrome now, see https://developer.chrome.com/blog/wasmgc/

It might make sense to re-alive this issue.

**Comment by Martin1994** [comment]
Why a Java AOT solution has to rely on the GC implementation of WASM? A Java GC should be runnable on WASM with low level memory management API, and on the other side using WASM's GC will most likely lose the GC related APIs on Java side.

**Comment by tenaf0** [comment]
@Martin1994 From the article: 

> The C and Rust binaries could be anywhere from 6.1 K to 9.6 K depending on the various compiler flags, while the Java version is much smaller at only 2.3 K!

Compare it to the size of a C# application compiled to WASM, which is on the MB range, due to in part to having to also download on every site a complete GC/runtime implementation.

Given that Java can already be compiled to WASM (closure compiler, teawm), I believe that the WASM GC proposal’s API is more than fit for Java’s use case, and there are many advantages to using Graal, like polyglottism, size reduction (due to the closed world assumption, it can remove a bunch of classes that are never accessed), etc.

**Comment by sgammon** [comment]
In any case, this issue deserves a fresh look, because it opens GVM binaries up to new deployment targets, regardless of GC

(Consider users of --gc=epsilon, for example)

GC being shipped in several popular engines is a signal that WASM in general is maturing and stabilizing.

**Comment by boonyasukd** [comment]
FYI, Firefox 120 now enables Wasm GC by default (it was hidden under a flag for quite some time). I'd say the browser side is quite ready now. Kotlin team has been demoing/optimizing their work on Wasm GC since march, it'd be great if Graal supports it also.

https://www.mozilla.org/en-US/firefox/120.0/releasenotes/
https://twitter.com/bashorov/status/1716891482487988630

**Comment by ericvergnaud** [comment]
WASM GC is available in V8 (Chrome, Deno and soon Node).
It'd be very disappointing to not see it rapidly coming to GraalVM

**Comment by a-cosmic-jaw** [comment]
Hello? Where are we on this?

**Comment by axel22** [comment]
We are working on and have an in-progress implementation that outputs Wasm (with the WasmGC spec).

It would be really helpful for us if people could explain their use-cases for having a WasmGC backend.
What sort of deployments scenarios do you foresee or have need for (e.g. do you need to target browser applications, or run Wasm modules in standalone mode with WASI)?

Thank you!

**Comment by gocursor** [comment]
> It would be really helpful for us if people could explain their use-cases for having a WasmGC backend. What sort of deployments scenarios do you foresee or have need for (e.g. do you need to target browser applications, or run Wasm modules in standalone mode with WASI)?

Primarily browser apps (to share the code between frontend and backend and to leverage the existing Java skills).

**Comment by CedNaru** [comment]
Hello, I am one of the maintainers of a Kotlin-Java binding for the Godot game engine (https://github.com/utopia-rise/godot-kotlin-jvm)
Godot supports exports for the main 3 desktop OS, Android, iOS and wasm. Wasm is currently the only export we don't support because we lack a simple way to run a JRE inside the browser. 
We already support the use of native images (our only way to run it on iOS currently) and a wasm target would be welcomed to complete our Godot support. 
An alternative would be to use Kotlin/Native and Kotlin/Wasm instead of native images, but Godot being a C++ engine, we rely heavily on interop. Using different backends for Kotlin would require us to implement a specific interop layer for each of them. It would be much simpler for us if we could just keep using a regular JRE + JNI context for all of them.

**Comment by agarciadom** [comment]
I am one of the developers of the [Eclipse Epsilon](https://eclipse.dev/epsilon/) project. We provide a family of mini-languages for managing object-oriented system models (e.g. validating them, generating code from them, and so on). They are all interpreted languages, running on top of Java.

We started as being Eclipse-centric (it was the shiny new thing back in the mid 00s), but we are seeing more and more people interested in doing modelling from their browsers (e.g. from low-code development platforms). In fact, to make it easier for people to try out our languages, we've created a web-based [Playground](https://eclipse.dev/epsilon/playground/) for the Epsilon languages.

The annoying part is that, since we can't run Java code on the browser, executing the code in the Playground involves spinning up a Google Cloud Function that runs the script, which is 1) costly and 2) slow. We'd very much like to use something like the Graal integration in Micronaut to package up our interpreter as a module that we could use from the Playground.

It'd also open other opportunities - for instance, we have the start of a language server for Epsilon, which is also written in Java. With a WasmGC export, we could look into running the language server from Node.js, which is one of the default runtimes supported by VS Code.

**Comment by lemmy** [comment]
> The annoying part is that, since we can't run Java code on the browser, executing the code in the Playground involves spinning up a Google Cloud Function that runs the script, which is 1) costly and 2) slow. We'd very much like to use something like the Graal integration in Micronaut to package up our interpreter as a module that we could use from the Playground

https://github.com/tlaplus/tlaplus has a pretty similar use case.

**Comment by gkresic** [comment]
> It would be really helpful for us if people could explain their use-cases for having a WasmGC backend.

Running Java code in browser environments would be our primary target, but WASI (although still in its infancy) should be ignored too. However, important feature in both cases would be as-good-as-possible support for integrations with JavaScript (in browser) and other WASM modules (especially ones written in languages other than Java).

**Comment by simlei** [comment]
Everything that has to do with compiling low-dependency Scala 3 programs to
WASM, basically. It needs GC support for that.

**Comment by lisonge** [comment]
> It would be really helpful for us if people could explain their use-cases for having a WasmGC backend.

I need to run Java regular expressions in the browser with the same behavior as on the JVM. However, the native JavaScript RegExp and Java's Pattern exhibit inconsistent behavior. For example, JavaScript does not allow embedded flags in regular expressions, whereas Java does.

It seems the only solution is to compile Java's Pattern to WebAssembly (wasm) for JavaScript to call.

My current solution is to use Kotlin Multiplatform to compile Kotlin's kotlin.text.Regex to wasm for JavaScript to call.

<https://github.com/gkd-kit/gkd/blob/v1.7.3/wasm_matches/src/commonMain/kotlin/li/songe/matches/toMatches.kt>

<https://github.com/gkd-kit/inspect/blob/f8b58ae6c34cf9559f148b566ddaadb0d3138cd4/src/utils/selector.ts#L14>

if Graal could support compiling Java to WebAssembly (wasm) for JavaScript to call, then more Java libraries could be run directly in the browser, eliminating the need to manually re-implement them in JavaScript

For example, I recently wanted to port the functionality of converting SVG to XML from Android Studio to the browser. The source code is available at [Svg2Vector.java](https://cs.android.com/android-studio/platform/tools/base/+/mirror-goog-studio-main:sdk-common/src/main/java/com/android/ide/common/vectordrawable/Svg2Vector.java). If these Java code could be directly compiled to WebAssembly (wasm), it would eliminate the need to manually re-implement it in JavaScript.

**Comment by bashor** [comment]
@lisonge, you can convert [Svg2Vector.java](https://cs.android.com/android-studio/platform/tools/base/+/mirror-goog-studio-main:sdk-common/src/main/java/com/android/ide/common/vectordrawable/Svg2Vector.java) to Kotlin 🫣 and continue using Kotlin Multiplatform 😉

**Comment by lisonge** [comment]
> @lisonge, you can convert [Svg2Vector.java](https://cs.android.com/android-studio/platform/tools/base/+/mirror-goog-studio-main:sdk-common/src/main/java/com/android/ide/common/vectordrawable/Svg2Vector.java) to Kotlin 🫣 and continue using Kotlin Multiplatform 😉

@bashor  The third-party library `org.w3c.dom` that it uses cannot be used with Kotlin Multiplatform

**Comment by sgammon** [comment]
@axel22 our use case is mostly server side, similar to Node. we want to support WASI as a server side target, especially on platforms like Cloudflare Workers.

many isolate-based hosting platforms can only accept WASM or JavaScript. compiling to WASM is therefore a fundamental need unless V8 begins accepting other binary types (unlikely).

this would also lower the cost floor for hosting a GraalVM application significantly. because WASM is not supported yet, one must resort to a container or a full VM, or even bare metal, all of which have much higher cost floors than isolate-based hosting.

**Comment by rdobrik** [comment]
We are working on Java stack for Internet Computer blockchain platform https://internetcomputer.org/. Currently developing Java Agent libraries but there is a demand for Java based backend smart contracts (canisters). Internet Computer backend is based on WASM and currently supports different languages like Rust, C++, Python and Javascript. While IO access will be quite restrictive, we will need to support Java reflections and annotations to be able serialize and deserialize Java types (like Jackson or Gson) to internal ICP data format. We already have code for that. Also support libraries like Bouncy Castle for encryption and decryption.

**Comment by nilslice** [comment]
This is really exciting! We'd love to be able to support a Java PDK for [Extism](https://github.com/extism/extism), which would mean supporting a Java library that calls some external functions (some static, like wasip1), and some created by a Host environment, linking them dynamically. 

Extism environments are embedded within other applications including things that run on the server, browser, IoT, and mobile.

**Comment by ningpengtao-coder** [comment]
I want to run my own swing application on wasm.

**Comment by Sintrastes** [comment]
Very interested in WASI / WIT support to be able have the graal ecosystem target a custom plugin ecosystem.

Basically the goal is to allow for third-parties to write a plugin in whatever language they want, so having the graal ecosystem as an option would be a huge win.

**Comment by nilslice** [comment]
@sintrastes - curious if you've come across https://extism.org, which makes all of that possible today across many more languages than WIT can generate. 

we don't have support for GraalWasm yet, but would be open to collaborating!

we also have a similar binding generator toolchain with support for Go, Rust, TypeScript, Python, C#, C++ and Zig (with more coming!). we release all of this freely and under permissive licenses, learn more here: https://docs.xtp.dylibso.com/docs/concepts/xtp-schema

**Comment by Sintrastes** [comment]
@nilslice - Interesting! No, I was not familiar with this previously.

In terms of the IDL, are there plans to add support for things like sum types, like in WIT?

**Comment by nilslice** [comment]
@Sintrastes - great question, and probably not in the near term, but it's not out of the question. 

XTP schema takes a very pragmatic approach to try and bring wasm to every app; vs try to make all apps wasm. So we take lots of care to not introduce features that are incompatible or create a non-idiomatic mess of generated code. 

Sum types could be done, but they don't exist in all the languages we support - so we'd have to hand over a wonky Frankenstein combination of structs/methods/enums/states etc. to our users and we have a hard time shipping things that don't work great. 

We also stick closely to OpenAPI spec, given that a local in-process plugin is basically just a bi-directional API call, and there's a LOT of OpenAPI based systems out there. 

I don't mean to take over the discussion here, so please feel free to join us on Discord: https://extism.org/discord

**Comment by ningpengtao-coder** [comment]
https://github.com/stackblitz/bolt.new has opened up a low-code tool for generating llm.
Using stackblitz's private webcontainer technology, it simulates the nodejs environment and runtime in the browser to support running front-end projects in the browser.
This fork of https://github.com/stackblitz/bolt.new plans to use open source components instead of webcontainer.
Open source components:
https://github.com/wasmerio/wasmer
https://github.com/Sandpack/nodebox-runtime
https://github.com/leaningtech/webvm
https://github.com/coleam00/bolt.new-any-llm/issues/229

Graal can introduce more runtime support, such as GraalPython, GraalJS, Espresso (Java), etc., and convert these runtimes into wasm.

My personal experience using graal:
I use graal to package swing programs into native and run them on desktop and mobile. But compared to browsers, it is very difficult to distribute programs on desktop and mobile.
From practical experience, small programs, PWA and web are commonly used now, which are easier to distribute. These are areas where wasm is more likely to exert its strength.

And now is the era of AI, and entrepreneurial technology selection is more inclined to languages ​​such as JS and Python. If wasm can be supported, Graal can be more organically combined with AI, and Graal will have a wider range of usage scenarios.

**Comment by Frontrider** [comment]
> And now is the era of AI, and entrepreneurial technology selection is more inclined to languages ​​such as JS and Python. If wasm can be supported, Graal can be more organically combined with AI, and Graal will have a wider range of usage scenarios.

Graal can get your jvm code to interact with python code and libraries directly. That is not the problem, that topic is not really a factor.

A more important entrepreneurial tech is that webassembly is cloud native, you can run it with docker directly.

**Comment by ningpengtao-coder** [comment]
@Frontrider This only satisfies running Python on the Java backend and running WASM in container runtimes (WasmEdge). It cannot run in browsers.

The following is a commercial product that can run Java applications in browsers, including Swing, Applet, and extensive framework support: https://cheerpj.com/

**Comment by ningpengtao-coder** [comment]
@ceisserer Thank you for your response. It's possible that CheerpJ does not support WebGPU very well? GraalVM's native-image compiles Java code into native code through the compiler toolchain, and it can also support WASM through the compiler toolchain. I am very hopeful that Java GUI technology can be brought back to the browser through GraalVM.

**Comment by Frontrider** [comment]
> @ceisserer Thank you for your response. It's possible that CheerpJ does not support WebGPU very well? GraalVM's native-image compiles Java code into native code through the compiler toolchain, and it can also support WASM through the compiler toolchain. I am very hopeful that Java GUI technology can be brought back to the browser through GraalVM.

This is already proven to be feasible, as this is basically what Kotlin Compose does for the web. Except it uses their own compiler.

I'm also reading some articles on SPA-s, and some of the things those are used for are literally better to do even with swing. When you have a lot of data to display, and the dom tree becomes way too convoluted.
The point is that it will come back as soon as the tech is feasible.

**Comment by mipastgt** [comment]
Now that Wasm GC has arrived in all major browsers (Chrome, Firefox and even Safari) is there any chance for some fresh wind in this feature request?

**Comment by mipastgt** [comment]
Did my wishes finally come true?
[The Future of Write Once, Run Anywhere: From Java to WebAssembly](https://2025.wasm.io/sessions/the-future-of-write-once-run-anywhere-from-java-to-webassembly/)

**Comment by x0wllaar** [comment]
Well, it exists now: https://graalvm.github.io/graalvm-demos/native-image/wasm-javac/

But there's still no official announcement of anything

**Comment by mipastgt** [comment]
More than one week after WASM I/O now and still no further info? :-(

**Comment by sideeffffect** [comment]
https://www.youtube.com/watch?v=Z2SWSIThHXY

**Comment by aochagavia** [comment]
The YouTube video linked above is showcasing a demo from [this repository](https://github.com/graalvm/graalvm-demos/tree/6b446b3c749705872f9e1cce5f5834981a8a5296/native-image/wasm-javac), which has instructions on how to compile your own code to wasm. There seems to be no additional documentation whatsoever about the feature, which I guess is expected since it's only available in early access builds.

**Comment by mipastgt** [comment]
No, it has just been officially release with GraalVM 25 but there still isn't much you can do with it.

**Comment by aochagavia** [comment]
Ah, thanks for the clarification. On a related note, I was trying the demo's command (using native-image with the `--tool:svm-wasm` option) and got the following error:

```
Unknown name in option specification: tool:svm-wasm
```

Did this ship in the release under a different option? I'm testing graal CE 25.0.0 on Windows, downloaded from the [official release](https://github.com/graalvm/graalvm-ce-builds/releases/).

**Comment by mipastgt** [comment]
It's in the Oracle release and not in the community edition. [Oracle GraalVM download](https://www.graalvm.org/downloads/)

**Comment by gkresic** [comment]
Also, note that [README.md](https://github.com/graalvm/graalvm-demos/blob/master/native-image/wasm-javac/README.md) from the `master` branch states _"An Early Access Build of Oracle GraalVM for JDK **26** or later"_ as a requirement.

So, I guess it's still not ready within GraalVM 25. However, it would be nice to have some official feedback on this, as it's obviously a much-awaited feature.

**Comment by mipastgt** [comment]
Nobody said it's ready. Actually it is not, but you don't need a pre-release anymore do use whatever is already available. Of course you will still find the latest features in the latest pre-release as always. I use the latest Oracle GraalVM 25 release and it does support the `tool:svm-wasm` option. Currently it can only generate main programs and no libraries. See: https://graalvm.slack.com/archives/CN9KSFB40/p1758097907532299

**Comment by fornof** [comment]
Does the current build of Oracle GraalVM support sending strings? GPT said it didn't and I needed to make my own memory allocator and pointer to use a string pass in to a function in java from javascript.

**Comment by fornof** [comment]
Are strings available on WASM to pass from js to Java? 
GPT said I needed to convert string , then malloc and use a pointer for it.

**Comment by fniephaus** [comment]
@fornof yes, there is [`JSString`](https://www.graalvm.org/sdk/javadoc/org/graalvm/webimage/api/JSString.html) which is, for example, [used in a few places in the javac demo](https://github.com/graalvm/graalvm-demos/blob/87b9c0a4efc880348e43462a7f0f26fd53d17a0b/web-image/javac/src/main/java/com/example/WebMain.java#L26).
